src/runtime_src/core/pcie/driver/linux/xocl/xocl_subdev.c

```diff
@@ -1301,36 +1311,61 @@ int xocl_subdev_create_vsec_devs(xdev_handle_t xdev)
 	ret = xocl_subdev_vsec(xdev, XOCL_VSEC_FLASH_CONTROLER, &bar, &offset,
 		&vtype);
 	if (!ret) {
-		struct xocl_subdev_info subdev_info = XOCL_DEVINFO_FLASH_VSEC;
+		struct xocl_subdev_info subdev_info =
+		    XOCL_DEVINFO_FLASH_VSEC;
+
+		switch (vtype) {
+		case XOCL_VSEC_FLASH_TYPE_QSPI:
+			memcpy(((struct xocl_flash_privdata *)
+			    (subdev_info.priv_data))->flash_type,
+			    FLASH_TYPE_QSPIPS, strlen(FLASH_TYPE_QSPIPS));
+			/* default is FLASH_TYPE_SPI, thus pass through. */
+		case XOCL_VSEC_FLASH_TYPE_SPI_IP:
+		case XOCL_VSEC_FLASH_TYPE_SPI_REG:
+			xocl_xdev_info(xdev,
+			    "VSEC FLASH RES Start 0x%llx, bar %d, type 0x%x",
+			    offset, bar, vtype);
+
+			ret = xocl_subdev_create_vsec_impl(xdev, &subdev_info,
+			    offset, bar);
 
-		xocl_xdev_info(xdev,
-			"Vendor Specific FLASH RES Start 0x%llx, bar %d",
-			 offset, bar);
-		subdev_info.res[0].start = offset;
-		subdev_info.res[0].end = offset + 0xfff;
-		subdev_info.bar_idx[0] = bar;
-		if (vtype == 0x2)
-                        memcpy(((struct xocl_flash_privdata *)
-				(subdev_info.priv_data))->flash_type,
-				FLASH_TYPE_QSPIPS, strlen(FLASH_TYPE_QSPIPS));
-
-		ret = xocl_subdev_create(xdev, &subdev_info);
-		if (ret)
-			return ret;
+			if (ret)
+				return ret;
+			break;
+		case XOCL_VSEC_FLASH_TYPE_VERSAL:
+			xocl_xdev_info(xdev,
+			    "VSEC VERSAL FLASH RES Start 0x%llx, bar %d",
+			    offset, bar);
+
+			/* set devinfo to xfer versal */
+			subdev_info.id = XOCL_SUBDEV_XFER_VERSAL;
+			subdev_info.name = XOCL_XFER_VERSAL;
+			subdev_info.res[0].name = XOCL_XFER_VERSAL;
+			memcpy(((struct xocl_flash_privdata *)
+			    (subdev_info.priv_data))->flash_type,
+			    FLASH_TYPE_OSPI_VERSAL, strlen(FLASH_TYPE_OSPI_VERSAL));
+
+			ret = xocl_subdev_create_vsec_impl(xdev, &subdev_info,
+				offset, bar);
+
+			if (ret)
+				return ret;
+			break;
+		default:
+			xocl_xdev_info(xdev, "Unsupport flash type 0x%x", vtype);
```

Comment:
it will be better to add a break even in default (default does not have to be the last case.)